llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp

```diff
@@ -1032,6 +1033,24 @@ static bool processSExt(SExtInst *SDI, LazyValueInfo *LVI) {
   return true;
 }
 
+static bool processZExt(ZExtInst *SDI, LazyValueInfo *LVI) {
+  if (SDI->getType()->isVectorTy())
+    return false;
+
+  if (SDI->hasNonNeg())
+    return false;
+
+  const Use &Base = SDI->getOperandUse(0);
+  if (!LVI->getConstantRangeAtUse(Base, /*UndefAllowed*/ false)
```

Comment:
Is there a test case covering at-use?

Reply:
IIUC the non-loop `@test3` covers at-use.

Reply:
I don't think it does. It has a dominating condition. I would expect something like a select condition for at-use.